stream_framer.go

```diff
@@ -57,14 +61,20 @@ func (f *streamFramer) HasFramesForRetransmission() bool {
 	return len(f.retransmissionQueue) > 0
 }
 
-// TODO: don't need to ask the crypto stream here, just record this information in AddActiveStream
-func (f *streamFramer) HasCryptoStreamFrame() bool {
-	return f.cryptoStream.hasDataForWriting()
+func (f *streamFramer) HasCryptoStreamData() bool {
+	f.streamQueueMutex.Lock()
+	hasCryptoStreamData := f.hasCryptoStreamData
+	f.streamQueueMutex.Unlock()
+	return hasCryptoStreamData
 }
 
-// TODO(lclemente): This is somewhat duplicate with the normal path for generating frames.
 func (f *streamFramer) PopCryptoStreamFrame(maxLen protocol.ByteCount) *wire.StreamFrame {
-	frame, _ := f.cryptoStream.popStreamFrame(maxLen)
+	frame, hasMoreData := f.cryptoStream.popStreamFrame(maxLen)
+	if !hasMoreData {
```

Comment:
I think there's a race here and you have to lock the mutex before popping. What happens if a Write() and an AddActiveStream() happen while in this line?

This may also apply to the non-crypto streams path – GH doesn't show that here.

Reply:
Good catch. The non-crypto stream code path should be fine, the mutex is locked in `maybePopNormalFrames` from beginning to end.